relay-server/src/actors/upstream.rs

```diff
@@ -881,69 +872,108 @@ impl Entry {
 /// is synchronous and managed by the [`UpstreamBroker`].
 #[derive(Debug)]
 struct UpstreamQueue {
+    /// High priority queue.
     high: VecDeque<Entry>,
+    /// Low priority queue.
     low: VecDeque<Entry>,
+    /// Low priority retry queue.
+    retry_high: VecDeque<Entry>,
+    /// Low priority retry queue.
+    retry_low: VecDeque<Entry>,
+    /// Retries should not be dequeued before this instant.
+    ///
+    /// This retry increments by the constant `retry_after` instead of backoff,
+    /// since it only kicks in for a short period of time before Relay gets into
+    /// network outage mode (see [`IsNetworkOutage`]).
+    next_retry: Instant,
+    /// Time to wait before retrying another request from the retry queue.
+    retry_after: Duration,
 }
 
 impl UpstreamQueue {
     /// Creates an empty upstream queue.
-    pub fn new() -> Self {
+    pub fn new(retry_after: Duration) -> Self {
         Self {
             high: VecDeque::new(),
             low: VecDeque::new(),
+            retry_high: VecDeque::new(),
+            retry_low: VecDeque::new(),
+            next_retry: Instant::now(),
+            retry_after,
         }
     }
 
     /// Returns the number of entries in the queue.
     pub fn len(&self) -> usize {
-        self.high.len() + self.low.len()
+        self.high.len() + self.low.len() + self.retry_high.len() + self.retry_low.len()
     }
 
-    /// Puts an entry into the queue at the given position.
-    fn put(&mut self, entry: Entry, position: EnqueuePosition) {
+    /// Places an entry at the back of the queue.
+    ///
+    /// Since entries are dequeued in FIFO order, this entry will be dequeued
+    /// last within its priority class; see
+    /// [`dequeue`][`UpstreamQueue::dequeue`] for more details.
+    pub fn enqueue(&mut self, entry: Entry) {
         let priority = entry.request.priority();
-
-        // We can ignore send errors here. Once the channel closes, we drop all incoming requests
-        // here. The receiving end of the request will be notified of the drop if they are waiting
-        // for it.
-        let queue = match priority {
-            RequestPriority::High => &mut self.high,
-            RequestPriority::Low => &mut self.low,
-        };
-
-        match position {
-            EnqueuePosition::Front => queue.push_front(entry),
-            EnqueuePosition::Back => queue.push_back(entry),
+        match priority {
+            RequestPriority::High => self.high.push_back(entry),
+            RequestPriority::Low => self.low.push_back(entry),
         }
-
         relay_statsd::metric!(
             histogram(RelayHistograms::UpstreamMessageQueueSize) = self.len() as u64,
             priority = priority.name(),
+            attempt = "first"
```

Comment:
I think observing how many attempts are coming from retries can be useful in certain scenarios.